tests/services/illumina_services/cleaning_services/test_clean_runs_service.py

```diff
@@ -0,0 +1,439 @@
+def test_can_sequencing_run_be_deleted_no_spring_with_fastq(
+    illumina_clean_service_can_be_removed: IlluminaCleanRunsService,
+):
+    """Test that a sequencing run can be deleted when it has fastq files but no spring files."""
+    # GIVEN a sequencing run that can be deleted
+
+    with mock.patch(
+        "cg.services.illumina_services.cleaning_services.clean_runs_service.IlluminaCleanRunsService.is_directory_older_than_21_days",
+        return_value=True,
+    ):
+        with mock.patch(
+            "cg.services.illumina_services.cleaning_services.clean_runs_service.IlluminaCleanRunsService.has_spring_meta_data_files_for_samples_in_housekeeper",
+            return_value=False,
+        ):
+            # WHEN checking that the sequencing run can be deleted
+            can_be_deleted: bool = (
+                illumina_clean_service_can_be_removed.can_run_directory_be_deleted()
+            )
+
+    # THEN the check whether the sequencing run can be deleted returns True
+    assert can_be_deleted
+
+
+def test_can_sequencing_run_be_deleted_spring_no_fastq(
+    illumina_clean_service_can_be_removed: IlluminaCleanRunsService,
+):
+    """Test that a sequencing run can be deleted when it has spring files but no fastq files."""
+    # GIVEN a sequencing run that can be deleted
+
+    with mock.patch(
+        "cg.services.illumina_services.cleaning_services.clean_runs_service.IlluminaCleanRunsService.is_directory_older_than_21_days",
+        return_value=True,
+    ):
+        with mock.patch(
+            "cg.services.illumina_services.cleaning_services.clean_runs_service.IlluminaCleanRunsService.has_fastq_files_for_samples_in_housekeeper",
+            return_value=False,
+        ):
+            # WHEN checking that the sequencing run can be deleted
+            can_be_deleted: bool = (
+                illumina_clean_service_can_be_removed.can_run_directory_be_deleted()
+            )
+
+    # THEN the check whether the sequencing run can be deleted returns True
+    assert can_be_deleted
+
+
+def test_can_sequencing_run_be_deleted_no_spring_no_fastq(
+    illumina_clean_service_can_be_removed: IlluminaCleanRunsService,
+):
+    """Test that a sequencing run can not be deleted when it has no spring files and no fastq files."""
+    # GIVEN a sequencing run that can be deleted
+
+    with mock.patch(
+        "cg.services.illumina_services.cleaning_services.clean_runs_service.IlluminaCleanRunsService.is_directory_older_than_21_days",
+        return_value=True,
+    ):
+        with mock.patch(
+            "cg.services.illumina_services.cleaning_services.clean_runs_service.IlluminaCleanRunsService.has_fastq_files_for_samples_in_housekeeper",
+            return_value=False,
+        ):
+            with mock.patch(
+                "cg.services.illumina_services.cleaning_services.clean_runs_service.IlluminaCleanRunsService.has_spring_meta_data_files_for_samples_in_housekeeper",
+                return_value=False,
+            ):
```

Comment:
there is no need to recursively add contexts, you can add one patch after another separated by a comma forming only one patching environment
```suggestion
    with mock.patch(  "cg.services.illumina_services.cleaning_services.clean_runs_service.IlluminaCleanRunsService.is_directory_older_than_21_days",
        return_value=True,
    ), with mock.patch(     "cg.services.illumina_services.cleaning_services.clean_runs_service.IlluminaCleanRunsService.has_fastq_files_for_samples_in_housekeeper",
        return_value=False,
    ), with mock.patch(            "cg.services.illumina_services.cleaning_services.clean_runs_service.IlluminaCleanRunsService.has_spring_meta_data_files_for_samples_in_housekeeper",
        return_value=False,
    ):
```